pkg/interop/client.go

```diff
@@ -0,0 +1,136 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interop
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+const (
+	// loRaAllianceDomain is the domain of LoRa Alliance.
+	loRaAllianceDomain = "lora-alliance.org"
+
+	// loRaAllianceJoinEUIDomain is the LoRa Alliance domain used for JoinEUI resolution.
+	loRaAllianceJoinEUIDomain = "joineuis." + loRaAllianceDomain
+
+	// loRaAllianceNetIDDomain is the LoRa Alliance domain used for NetID resolution.
+	loRaAllianceNetIDDomain = "netids." + loRaAllianceDomain
+)
+
+func LoRaAllianceJoinServerFQDN(eui types.EUI64) string {
+	o := []rune(strings.ToLower(eui.String()))
+	r := make([]rune, 2*len(o)-1)
+	// Inspired by https://github.com/golang/example/blob/46695d81d1fae905a270fb7db8a4d11a334562fe/stringutil/reverse.go#L20-L27
+	for i, j := 0, len(o)-1; i < len(o)/2; i, j = i+1, j-1 {
+		r[i+i], r[j+j] = o[j], o[i]
+		r[i+i+1] = '.'
+		r[j+j-1] = '.'
+	}
+	return fmt.Sprintf("%s.%s", string(r), loRaAllianceJoinEUIDomain)
+}
+
+func HandleJoinRequest(ctx context.Context, tlsConfig *tls.Config, netID types.NetID, joinEUI, devEUI types.EUI64, req *ttnpb.JoinRequest, fqdn string) (*ttnpb.JoinResponse, error) {
+	dlSettings, err := lorawan.MarshalDLSettings(req.DownlinkSettings)
+	if err != nil {
+		return nil, ErrJoinReq.WithCause(err)
+	}
+
+	var cfList []byte
+	if req.CFList != nil {
+		cfList, err = lorawan.MarshalCFList(*req.CFList)
+		if err != nil {
+			return nil, ErrJoinReq.WithCause(err)
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	if err := json.NewEncoder(buf).Encode(&JoinReq{
+		NsJsMessageHeader: NsJsMessageHeader{
+			MessageHeader: MessageHeader{
+				ProtocolVersion: "1.0",
+				MessageType:     MessageTypeJoinReq,
+				// TODO: Add TransactionID if necessary
+				// TODO: Add SenderToken if necessary
+			},
+			SenderID:   NetID(netID),
+			ReceiverID: EUI64(joinEUI),
+			SenderNSID: NetID(netID),
+		},
+		MACVersion: MACVersion(req.SelectedMACVersion),
+		PHYPayload: Buffer(req.RawPayload),
+		DevEUI:     EUI64(devEUI),
+		DevAddr:    DevAddr(req.DevAddr),
+		DLSettings: Buffer(dlSettings),
+		RxDelay:    req.RxDelay,
+		CFList:     Buffer(cfList),
+	}); err != nil {
+		return nil, err
+	}
+
+	scheme := "http"
+	var cl http.Client
+	if tlsConfig != nil {
+		scheme = "https"
+		cl.Transport = &http.Transport{
+			TLSClientConfig: tlsConfig,
+		}
+	}
+
+	res, err := cl.Post(fmt.Sprintf("%s://%s", scheme, fqdn), "application/json", buf)
```

Comment:
We need JS options through config here for path and port

Reply:
See https://github.com/thethingsnetwork/lorawan-stack/issues/4#issuecomment-504048663

Reply:
So if I refactor the `fqdn string` parameter to `url string` this is fixed, right?

Reply:
So then also `LoRaAllianceJoinServerFQDN` turns into `LoRaAllianceJoinServerURL` and takes e.g. `types.EUI64{0x70, 0xb3, 0xd5, 0x7e, 0xd0, 0x00, 0x00, 0x00}` to `"https://0.0.0.0.0.0.0.d.e.7.5.d.3.b.0.7.joineuis.lora-alliance.org"` ?

Reply:
Note, that nothing prevents you from passing `whatever.foo.bar:1234/asd` as the PRL